src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -1263,6 +1405,81 @@ protected virtual async Task<DateTime> OnRenewLockAsync(string lockToken)
             return lockedUntilUtc;
         }
 
+        protected virtual async Task<DateTime> OnRenewLocksAsync(IEnumerable<string> lockTokens)
```

Comment:
>OnRenewLocksAsync [](start = 47, length = 17)

You should test these. I am not sure if this works from service side.

Reply:
What about cases when in a batch of 100 messages, 50 are already consumed (or completed) and you are renewing all 100. ?

---
In reply to: [255765849](https://github.com/Azure/azure-service-bus-dotnet/pull/636#discussion_r255765849) [](ancestors = 255765849)

Reply:
Do not proceed till you've tested these base foundations as I feel you might get blocked somehwere. 

---
In reply to: [255766004](https://github.com/Azure/azure-service-bus-dotnet/pull/636#discussion_r255766004) [](ancestors = 255766004,255765849)